### PR TITLE
support of float dtypes for draw_segmentation_masks

### PR DESCRIPTION
**Pull request by GsnMithra**
Fixes: #8138

Hey there! 

I've implemented support for draw_* methods to seamlessly handle both uint8 and float32 image types. The processed image will now be returned with the same data type as the input. Your feedback is always appreciated.

Thank you!

### CODE REVIEW

File: torchvision/utils.py [hunk 1]
```diff
@@ -282,8 +282,8 @@ def draw_segmentation_masks(
         _log_api_usage_once(draw_segmentation_masks)
     if not isinstance(image, torch.Tensor):
         raise TypeError(f"The image must be a tensor, got {type(image)}")
-    elif image.dtype != torch.uint8:
-        raise ValueError(f"The image dtype must be uint8, got {image.dtype}")
+    elif image.dtype not in {torch.uint8, torch.float32}:
+        raise ValueError(f"The image dtype must be uint8 or float32, got {image.dtype}")
```
Comment (NicolasHug) [inline torchvision/utils.py:286]: Here and in the other places, let's just check for `image.is_floating_point()` instead of checking specifically for `float32`. This way we can also support float64, etc.

File: torchvision/utils.py [hunk 2]
```diff
@@ -164,11 +164,11 @@ def draw_bounding_boxes(
 
     """
     Draws bounding boxes on given image.
-    The values of the input image should be uint8 between 0 and 255.
+    The image values should be in the uint8 or float32 that range between 0 and 255.
```
Comment (NicolasHug) [inline torchvision/utils.py:167]: Here and in the other places:

```suggestion
    The image values should be uint8 in [0, 255] or float in [0, 1]
```

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot for submitting this PR @GsnMithra

This is promising, but allowing support for float will require a little bit more work than just converting the dtype using `.to()`, since the convention for float images is that their valuae range is in [0, 1] instead of [0, 255].

Perhaps the easiest way to make progress here will be to focus on `draw_segmentation_masks` first, and to add a small unit test in https://github.com/pytorch/vision/blob/main/test/test_utils.py, that checks for equality of result for a given image with different dtypes. Something roughly like this:

```
from torchvision.transforms.v2.functional import to_dtype

img_uint8 = torch.randing(0, 256, (3, 100, 100), dtype=torch.uint8)
img_float = to_dtype(img_uint8, torch.float32, scale=True)

out_uint8 = draw_segmentation_mask(img_uint8, ...)
out_float = draw_segmentation_mask(img_float, ...)

torch.testing.assert_close(out_uint8, to_dtype(out_float, torch.uint8, scale=True))
```

Does that make sense?

File: torchvision/utils.py (hunk 1 above)
Comment (GsnMithra) [reply torchvision/utils.py:286]: Hey, thanks for replying, this will be fixed with the upcoming commits.

**Comment by GsnMithra** [comment]
Hey @NicolasHug 

Just dropping a note to mention that I've added some assertion checks for floating-point data types in the unit test method for draw_segmentation_mask() with the recent commit titled "test_draw_sementation_mask". I would love to get your thoughts on it.

Thank you for your time.

**Comment by GsnMithra** [comment]
Hello @NicolasHug 

I hope this message finds you well. I would like to extend my sincere apologies for any confusion caused by my recent actions. As a newcomer to this repository, I am still navigating the learning curve and may have made a mistake.

Specifically, as I attempted to write a unit test method for test_draw_segmentation_masks_dtypes(), I encountered an issue when referring to the draw_segmentation_masks() function. It appears that the library function lacks the float handler, which I have locally updated in the utils.py file. 

I genuinely appreciate your expertise and guidance in resolving this matter. I am eager to contribute to this repository and I am committed to learning and improving.

Thank you very much for your time and understanding.

**Comment by GsnMithra** [comment]
Hey @NicolasHug 

I would like to sincerely apologize for the mistakes I made in my previous contributions. I am still in the learning process and appreciate your guidance.

In the latest commit, I have included a new unit test called test_draw_segmentation_masks_dtypes, which has been implemented according to the suggestions provided earlier. I would greatly appreciate your thoughts and feedback on this addition.

Once again, I apologize for any inconvenience caused and thank you for taking the time to review my work.

File: test/test_utils.py [hunk 3]
```diff
@@ -246,6 +247,25 @@ def test_draw_segmentation_masks(colors, alpha, device):
...
+    out_uint8 = utils.draw_segmentation_masks(img_uint8, masks)
+
+    assert img_uint8 is not out_uint8
+    assert out_uint8.dtype == torch.uint8
+
+    img_float = to_dtype(img_uint8, torch.float, scale=True)
+    out_float = utils.draw_segmentation_masks(img_float, masks)
+
+    assert img_float is not out_float
+    assert out_float.is_floating_point ()
```
Comment (NicolasHug) [inline test/test_utils.py:265]:
```suggestion
    assert out_float.is_floating_point()
```

File: .gitignore [hunk 4]
```diff
@@ -42,3 +42,5 @@ xcuserdata/
 # direnv
 .direnv
 .envrc
+
+.DS_Store
```
Comment (NicolasHug) [inline .gitignore:46]: Let's revert these changes, we shouldn't need to modify this file here

```suggestion

```

File: test/test_utils.py [hunk 5]
```diff
@@ -7,8 +7,9 @@
 import numpy as np
 import pytest
 import torch
-import torchvision.transforms.functional as F
 import torchvision.utils as utils
+import torchvision.transforms.functional as F
+from torchvision.transforms.v2.functional import to_dtype
```
Comment (NicolasHug) [inline test/test_utils.py:12]: You'll need to re-order the imports, check out https://github.com/pytorch/vision/actions/runs/7214131414/job/19676695616?pr=8150, or run the pre-commit hooks locally (check our contributing instructions)

File: test/test_utils.py [hunk 6]
```diff
@@ -246,6 +247,25 @@ def test_draw_segmentation_masks(colors, alpha, device):
...
+    assert img_uint8 is not out_uint8
+    assert out_uint8.dtype == torch.uint8
+
+    img_float = to_dtype(img_uint8, torch.float, scale=True)
+    out_float = utils.draw_segmentation_masks(img_float, masks)
+
+    assert img_float is not out_float
+    assert out_float.is_floating_point ()
+
+    torch.testing.assert_close(out_uint8, to_dtype(out_float, torch.uint8, scale=True), rtol=0, atol=1.0)
```
Comment (NicolasHug) [inline test/test_utils.py:267]:
```suggestion
    torch.testing.assert_close(out_uint8, to_dtype(out_float, torch.uint8, scale=True), rtol=0, atol=1)
```

File: torchvision/utils.py [hunk 7]
```diff
@@ -297,15 +297,18 @@ def draw_segmentation_masks(
     if masks.shape[-2:] != image.shape[-2:]:
         raise ValueError("The image and the masks must have the same height and width")
 
+    original_dtype = image.dtype
+    if image.is_floating_point():
+        image = (image * 255).to(torch.uint8)
```
Comment (NicolasHug) [inline torchvision/utils.py:302]: Let's use `to_dtype` here instead, like you did in the test.

File: torchvision/utils.py [hunk 8]
```diff
@@ -315,7 +318,10 @@ def draw_segmentation_masks(
         img_to_draw[:, mask] = color[:, None]
 
     out = image * (1 - alpha) + img_to_draw * alpha
-    return out.to(out_dtype)
+    if original_dtype in {torch.float16, torch.float32, torch.float64}:
+        out = out.float() / 255.0
```
Comment (NicolasHug) [inline torchvision/utils.py:322]: Same: Let's use `to_dtype` here instead, like you did in the test.

File: torchvision/utils.py [hunk 9]
```diff
@@ -315,7 +318,10 @@ def draw_segmentation_masks(
         img_to_draw[:, mask] = color[:, None]
 
     out = image * (1 - alpha) + img_to_draw * alpha
-    return out.to(out_dtype)
+    if original_dtype in {torch.float16, torch.float32, torch.float64}:
```
Comment (NicolasHug) [inline torchvision/utils.py:321]: Check for `is_floating_point()` instead.

**Review by NicolasHug — COMMENTED** [review]
Thank you for the follow-up @GsnMithra, and no problems for the first few hicups.

I can see the new test is passing now, good job! In order to move forward with this PR, I would suggest the following:

- Address the comments I made below (it should be pretty easy)
- Remove the changes that were made to `draw_bounding_boxes()` and `draw_keypoints()`, and only keep the changes and tests relating to `draw_segmentation_masks()`. This way, we'll be able to merge this PR straight away, and then you can send 2 separate PR (one for boxes, one for keypoints), should you wish to.

Does that sound good to you?

File: torchvision/utils.py (hunk 9 above)
Comment (GsnMithra) [reply torchvision/utils.py:321]: Hey there! @NicolasHug 

We can use to_dtypes as mentioned above but there might be two issues with it:
1) `from torchvision.transforms.v2.functional import to_dtype` when imported normally (globally) gives us circular import due to the `_log_api_usage_once(to_dtype)` which is being imported in the to_dtype function, to overcome this we can either import in inside `draw_segmentation_masks()` method or we manually convert it ourselves.

2) When float dtype image is passed converting it back to float like so 
```
out = image * (1 - alpha) + img_to_draw * alpha
if original_image.is_floating_point():
    out = to_dtype(out, torch.float, scale=True)
```

is resulting in mismatched elements for the unit test, and is only working when we divide it with 255.0
```
out = image * (1 - alpha) + img_to_draw * alpha
if original_image.is_floating_point():
    out = to_dtype(out, torch.float) / 255.0
```

Thoughts on this?
Appreciate your time, thank you!

**Comment by GsnMithra** [comment]
Hello @NicolasHug 

I hope this message finds you well. I wanted to bring up a couple of points regarding the usage of `to_dtype` in the `torchvision.transforms.v2.functional` module, particularly within the `draw_segmentation_masks()` method.

Firstly, when importing `to_dtype` globally, it leads to a circular import issue due to the `_log_api_usage_once(to_dtype)` being utilized within the function. One solution could be importing it inside the `draw_segmentation_masks()` method, or alternatively, manually converting it.

Secondly, I've encountered a situation where, when a float dtype image is passed, converting it back to float using the following code results in mismatched elements for the unit test:

```
out = image * (1 - alpha) + img_to_draw * alpha
if original_image.is_floating_point():
    out = to_dtype(out, torch.float, scale=True)
```

It seems to work only when dividing by 255.0:

```
out = image * (1 - alpha) + img_to_draw * alpha
if original_image.is_floating_point():
    out = to_dtype(out, torch.float) / 255.0
```

I appreciate your insights and look forward to any further guidance or suggestions you may have.
Thank you for taking the time.

File: torchvision/utils.py (hunk 9 above)
Comment (GsnMithra) [reply torchvision/utils.py:321]: Hello @NicolasHug,

I hope this message finds you well. I wanted to seek your advice regarding the usage of the is_floating_point() method. In order to utilize it, one option is to pass in a tensor as follows:

`torch.is_floating_point(torch.as_tensor(0, dtype=original_dtype))`

Alternatively, we could clone the image `image.clone()` and check the cloned image as `clone_image.is_floating_point()`

I would appreciate your suggestion on the most appropriate approach or if there is any other recommended way to proceed in this context.

**Review by NicolasHug — APPROVED** [review]
Thanks a lot for the follow-up @GsnMithra !

You are right about the need for `/= 255`. It was because `out` was a float32 image in [0, 255] when the input was a float image. So even doing `dtype(scale=True)` would not work as expected, as no conversion was done (since the image was already float32).

I found a simpler way to handle all this by simply converting the color's dtype and scale, instead of converting the input image. I pushed the changes and also reverted some minor things from the other functions.

I'll merge this PR when green, thanks a ton for your work on this!

(Feel free to submit follow-up PRs for the other functions if you wish to. Just giving you a heads-up that I will only be able to review them starting next year in Jan, as I'll be on leaves from tomorrow.)

Thanks again @GsnMithra  !

**Comment by GsnMithra** [comment]
Hey @NicolasHug 
You're welcome, I'll be submitting PR's for the other functions also. 

Wishing you a wonderful break ;)
